### DISCUSSION TITLE
Generate new note using templater API from another plugin

### DISCUSSION THREAD

**Discussion by LeonardoGentile** [SilentVoid13/Templater#1391] [Help]
I'd want to create a new note using the templater command/api `Templater: Create new note from template` from outside templater. 

**Context**
I created some modules and scripts using https://github.com/polyipseity/obsidian-modules then I've created an almost black template that contains my user function entry point (defined as obsidian-modules)
So when templater generate a new note from this template, it executes my scripts that populate and move around the note. So far so good. 

Now I wanted to embed all this logic into a single plugin, that means: my plugin will call templater `create_new` and pass my almost empty template to it. This template contains a single entry points to functions and module defined into my plugin.

How would I do that? 

Is it as simple as passing my template path to `create_new` ? Also if my template contains user functions that are only stored in my plugin internals I don't think templater will be happy as user functions must be all defined under a certain path, isn't it? 

Any hint will be highly welcome 🙏

**Comment by LeonardoGentile** [comment]
Otherwise, I was wondering if I could extend the internal function / variable dynamically as described here  https://silentvoid13.github.io/Templater/internal-functions/contribute.html 

Meaning that instead of creating a pr and submitting it, would it be possible if my plugin could extends templater at run-time? 🤔

**Reply by McGlear** [reply]
What exactly are you trying to accomplish?
What do your current scripts do?

I use the modules plugin as well (because I wanted to import node packages like dedent) and programmatically create a string (well, a dedent-tagged Javascript template literal) containing my template that I then pass to `tp.file.create_new(my_template_string, filename, false, folder)`. That works without any problems.
What kind of functionality are you trying to gain by creating your own plugin? And why _shouldn't_ you be able to create whatever template you want and pass it to templater commands as arguments, whether you create that template within a personal plugin or a js module exported from the modules plugin?

**Reply by LeonardoGentile** [reply]
hello @McGlear thanks for your insight! 
Yes, short after my question I've realized I can pass a string to create_new, that would solve part of my problem. 
As it stands right now, I have a plain template that just call my entry point defined in obsidian-module. In there, I have a complex logic interfacing templater with metadata menu and presenting the user with several choices and popups, based on user choices it will generate specific notes from a unique generic templates.

Now I want to embed all this logic into a plugin that will act as a bridge between templater, metadata menu and few other plugins. 
The thing is that my entry point as it is right now, using obsidian-module, is equivalent to a user script, that mean I can pass the `tp` instance to my entry point functions and call all the templater functionalities from my modules. This work as it should. 

I want to reach the same functionalities by embedding all of this into a plugin. 
When I trigger my plugin main action it will call `create_new` but to generate the string to feed into `create_new` I need to pass `tp` to my main functions that generate the template string. 

I could create a simple templater template and in there call my entry point (so I could inject `tp`) but in this case my entry point is not a user function (it's not stored into templater user functions path) it's my plugin api, so this is the crux 🤔

I'm not sure I explained myself clearly, probably the solution is easier than I think but I can't see it from my point of view 🤷‍♂️

**Reply by McGlear** [reply]
I am still not sure if I understand you correctly, but I think you may want to look at how to [access a plugin's API-object from other scripts](https://forum.obsidian.md/t/js-using-the-template-tp-object-outside-of-templater-scripts-like-in-the-console-or-customjs/71485) and also at [defining another plugin as something akin to a dependency](https://forum.obsidian.md/t/how-to-access-other-plugins-as-dependencies/14469).

If that does not help, you could try to provide a minimal working example of your current setup and restate your questions based on that example.
